File: src/exchange/binance.py
"""Binance API client wrapper."""

import asyncio
from decimal import Decimal, ROUND_DOWN
from typing import Optional, Callable
import time

from binance import AsyncClient, BinanceSocketManager
from binance.exceptions import BinanceAPIException

from src.models.schemas import Order, OrderSide, OrderStatus, Trade, Balance
from src.utils.config import get_settings
from src.utils.logger import get_logger

logger = get_logger(__name__)


class BinanceClient:
    """Async Binance API client wrapper."""

    def __init__(self):
        self.settings = get_settings()
        self.client: Optional[AsyncClient] = None
        self.socket_manager: Optional[BinanceSocketManager] = None
        self._symbol_info: dict = {}
        self._price_callbacks: list[Callable] = []
        self._order_callbacks: list[Callable] = []
        self._running = False

    async def connect(self) -> None:
        """Initialize connection to Binance."""
        try:
            self.client = await AsyncClient.create(
                api_key=self.settings.binance_api_key,
                api_secret=self.settings.binance_api_secret,
                testnet=self.settings.binance_testnet,
            )
            self.socket_manager = BinanceSocketManager(self.client)

            # Load symbol info for the trading pair
            await self._load_symbol_info()

            logger.info(
                "binance_connected",
                testnet=self.settings.binance_testnet,
                trading_pair=self.settings.trading_pair,
            )
        except Exception as e:
            logger.error("binance_connection_failed", error=str(e))
            raise

    async def disconnect(self) -> None:
        """Close connection to Binance."""
        self._running = False
        if self.client:
            await self.client.close_connection()
            logger.info("binance_disconnected")

    async def _load_symbol_info(self) -> None:
        """Load trading pair information for precision."""
        if not self.client:
            return

        info = await self.client.get_symbol_info(self.settings.trading_pair)
        if info:
            self._symbol_info = {
                "symbol": info["symbol"],
                "base_asset": info["baseAsset"],
                "quote_asset": info["quoteAsset"],
                "price_precision": info["quotePrecision"],
                "quantity_precision": info["baseAssetPrecision"],
            }

            # Extract filters
            for f in info["filters"]:
                if f["filterType"] == "PRICE_FILTER":
                    self._symbol_info["tick_size"] = float(f["tickSize"])
                    self._symbol_info["min_price"] = float(f["minPrice"])
                    self._symbol_info["max_price"] = float(f["maxPrice"])
                elif f["filterType"] == "LOT_SIZE":
                    self._symbol_info["step_size"] = float(f["stepSize"])
                    self._symbol_info["min_qty"] = float(f["minQty"])
                    self._symbol_info["max_qty"] = float(f["maxQty"])
                elif f["filterType"] == "NOTIONAL":
                    self._symbol_info["min_notional"] = float(f.get("minNotional", 0))

            logger.debug("symbol_info_loaded", info=self._symbol_info)

    def round_price(self, price: float) -> float:
        """Round price to valid tick size."""
        if "tick_size" not in self._symbol_info:
            return price

        tick_size = Decimal(str(self._symbol_info["tick_size"]))
        price_decimal = Decimal(str(price))
        return float(price_decimal.quantize(tick_size, rounding=ROUND_DOWN))

    def round_quantity(self, quantity: float) -> float:
        """Round quantity to valid step size."""
        if "step_size" not in self._symbol_info:
            return quantity

        step_size = Decimal(str(self._symbol_info["step_size"]))
        qty_decimal = Decimal(str(quantity))
        return float(qty_decimal.quantize(step_size, rounding=ROUND_DOWN))

    async def get_current_price(self) -> float:
        """Get current price for trading pair."""
        if not self.client:
            raise RuntimeError("Client not connected")

        ticker = await self.client.get_symbol_ticker(symbol=self.settings.trading_pair)
        return float(ticker["price"])

    async def get_balances(self) -> list[Balance]:
        """Get account balances."""
        if not self.client:
            raise RuntimeError("Client not connected")

        account = await self.client.get_account()
        balances = []

        for b in account["balances"]:
            free = float(b["free"])
            locked = float(b["locked"])
            if free > 0 or locked > 0:
                balances.append(Balance(
                    asset=b["asset"],
                    free=free,
                    locked=locked,
                ))

        return balances

    async def get_balance(self, asset: str) -> Balance:
        """Get balance for specific asset."""
        if not self.client:
            raise RuntimeError("Client not connected")

        account = await self.client.get_account()

        for b in account["balances"]:
            if b["asset"] == asset:
                return Balance(
                    asset=b["asset"],
                    free=float(b["free"]),
                    locked=float(b["locked"]),
                )

        return Balance(asset=asset, free=0.0, locked=0.0)

    async def place_limit_order(
        self,
        side: OrderSide,
        price: float,
        quantity: float,
        grid_level: Optional[int] = None,
    ) -> Order:
        """Place a limit order."""
        if not self.client:
            raise RuntimeError("Client not connected")

        # Round to valid precision
        price = self.round_price(price)
        quantity = self.round_quantity(quantity)

        # Validate minimum notional
        notional = price * quantity
        min_notional = self._symbol_info.get("min_notional", 0)
        if notional < min_notional:
            raise ValueError(f"Order notional {notional} below minimum {min_notional}")

        try:
            client_order_id = f"grid_{grid_level}_{side.value}_{int(time.time() * 1000)}"

            result = await self.client.create_order(
                symbol=self.settings.trading_pair,
                side=side.value,
                type="LIMIT",
                timeInForce="GTC",
                price=str(price),
                quantity=str(quantity),
                newClientOrderId=client_order_id,
            )

            order = Order(
                order_id=str(result["orderId"]),
                client_order_id=result["clientOrderId"],
                trading_pair=result["symbol"],
                side=OrderSide(result["side"]),
                order_type=result["type"],
                price=float(result["price"]),
                quantity=float(result["origQty"]),
                filled_quantity=float(result["executedQty"]),
                status=OrderStatus(result["status"]),
                grid_level=grid_level,
            )

            logger.info(
                "order_placed",
                order_id=order.order_id,
                side=side.value,
                price=price,
                quantity=quantity,
                grid_level=grid_level,
            )

            return order

        except BinanceAPIException as e:
            logger.error(
                "order_failed",
                error=str(e),
                side=side.value,
                price=price,
                quantity=quantity,
            )
            raise

    async def cancel_order(self, order_id: str) -> bool:
        """Cancel an order."""
        if not self.client:
            raise RuntimeError("Client not connected")

        try:
            await self.client.cancel_order(
                symbol=self.settings.trading_pair,
                orderId=int(order_id),
            )
            logger.info("order_cancelled", order_id=order_id)
            return True
        except BinanceAPIException as e:
            logger.error("cancel_failed", order_id=order_id, error=str(e))
            return False

    async def cancel_all_orders(self) -> int:
        """Cancel all open orders for the trading pair."""
        if not self.client:
            raise RuntimeError("Client not connected")

        try:
            result = await self.client.cancel_open_orders(
                symbol=self.settings.trading_pair
            )
            count = len(result) if isinstance(result, list) else 0
            logger.info("all_orders_cancelled", count=count)
            return count
        except BinanceAPIException as e:
            logger.error("cancel_all_failed", error=str(e))
            return 0

    async def get_open_orders(self) -> list[Order]:
        """Get all open orders."""
        if not self.client:
            raise RuntimeError("Client not connected")

        orders = await self.client.get_open_orders(symbol=self.settings.trading_pair)

        return [
            Order(
                order_id=str(o["orderId"]),
                client_order_id=o["clientOrderId"],
                trading_pair=o["symbol"],
                side=OrderSide(o["side"]),
                order_type=o["type"],
                price=float(o["price"]),
                quantity=float(o["origQty"]),
                filled_quantity=float(o["executedQty"]),
                status=OrderStatus(o["status"]),
                grid_level=self._extract_grid_level(o["clientOrderId"]),
            )
            for o in orders
        ]

    async def get_order(self, order_id: str) -> Optional[Order]:
        """Get a specific order by ID."""
        if not self.client:
            raise RuntimeError("Client not connected")

        try:
            o = await self.client.get_order(
                symbol=self.settings.trading_pair,
                orderId=int(order_id),
            )

            return Order(
                order_id=str(o["orderId"]),
                client_order_id=o["clientOrderId"],
                trading_pair=o["symbol"],
                side=OrderSide(o["side"]),
                order_type=o["type"],
                price=float(o["price"]),
                quantity=float(o["origQty"]),
                filled_quantity=float(o["executedQty"]),
                status=OrderStatus(o["status"]),
                grid_level=self._extract_grid_level(o["clientOrderId"]),
            )
        except BinanceAPIException:
            return None

    async def get_recent_trades(self, limit: int = 50) -> list[Trade]:
        """Get recent trades for the account."""
        if not self.client:
            raise RuntimeError("Client not connected")

        trades = await self.client.get_my_trades(
            symbol=self.settings.trading_pair,
            limit=limit,
        )

        return [
            Trade(
                trade_id=str(t["id"]),
                order_id=str(t["orderId"]),
                trading_pair=t["symbol"],
                side=OrderSide.BUY if t["isBuyer"] else OrderSide.SELL,
                price=float(t["price"]),
                quantity=float(t["qty"]),
                commission=float(t["commission"]),
                commission_asset=t["commissionAsset"],
            )
            for t in trades
        ]

    def _extract_grid_level(self, client_order_id: str) -> Optional[int]:
        """Extract grid level from client order ID."""
        if client_order_id and client_order_id.startswith("grid_"):
            parts = client_order_id.split("_")
            if len(parts) >= 2:
                try:
                    return int(parts[1])
                except ValueError:
                    pass
        return None

    def on_price_update(self, callback: Callable) -> None:
        """Register callback for price updates."""
        self._price_callbacks.append(callback)

    def on_order_update(self, callback: Callable) -> None:
        """Register callback for order updates."""
        self._order_callbacks.append(callback)

    async def start_price_stream(self) -> None:
        """Start WebSocket stream for price updates."""
        if not self.socket_manager:
            raise RuntimeError("Socket manager not initialized")

        self._running = True

        async with self.socket_manager.symbol_ticker_socket(
            self.settings.trading_pair
        ) as stream:
            while self._running:
                try:
                    msg = await asyncio.wait_for(stream.recv(), timeout=30)
                    if msg and "c" in msg:
                        price = float(msg["c"])
                        for callback in self._price_callbacks:
                            await callback(price)
                except asyncio.TimeoutError:
                    continue
                except Exception as e:
                    logger.error("price_stream_error", error=str(e))
                    if self._running:
                        await asyncio.sleep(5)

    async def start_user_stream(self) -> None:
        """Start WebSocket stream for user updates (orders, trades)."""
        if not self.client or not self.socket_manager:
            raise RuntimeError("Client not initialized")

        self._running = True

        async with self.socket_manager.user_socket() as stream:
            while self._running:
                try:
                    msg = await asyncio.wait_for(stream.recv(), timeout=60)
                    if msg:
                        await self._handle_user_message(msg)
                except asyncio.TimeoutError:
                    continue
                except Exception as e:
                    logger.error("user_stream_error", error=str(e))
                    if self._running:
                        await asyncio.sleep(5)

    async def _handle_user_message(self, msg: dict) -> None:
        """Handle user stream messages."""
        event_type = msg.get("e")

        if event_type == "executionReport":
            order = Order(
                order_id=str(msg["i"]),
                client_order_id=msg["c"],
                trading_pair=msg["s"],
                side=OrderSide(msg["S"]),
                order_type=msg["o"],
                price=float(msg["p"]),
                quantity=float(msg["q"]),
                filled_quantity=float(msg["z"]),
                status=OrderStatus(msg["X"]),
                grid_level=self._extract_grid_level(msg["c"]),
            )

            for callback in self._order_callbacks:
                await callback(order)

            logger.debug(
                "order_update",
                order_id=order.order_id,
                status=order.status.value,
                filled=order.filled_quantity,
            )

File: src/exchange/binance.ts
import Binance from 'binance';
import { config } from '../utils/config.js';
import { createLogger } from '../utils/logger.js';
import type { Order, OrderSide, OrderStatus, Trade, Balance } from '../types/index.js';

const logger = createLogger('binance');

type PriceCallback = (price: number) => void;
type OrderCallback = (order: Order) => void;

// Simulated order for tracking in simulation mode
interface SimulatedOrder extends Order {
  isSimulated: true;
}

export class BinanceClient {
  private client: Binance.MainClient | null = null;
  private wsClient: Binance.WebsocketClient | null = null;
  private symbolInfo: {
    tickSize?: number;
    stepSize?: number;
    minNotional?: number;
  } = {};
  private priceCallbacks: PriceCallback[] = [];
  private orderCallbacks: OrderCallback[] = [];
  private isConnected = false;

  // Simulation state
  private simulatedOrders: Map<string, SimulatedOrder> = new Map();
  private simulatedOrderIdCounter = 1000000;
  private simulatedBalances: Map<string, Balance> = new Map();
  private _lastPrice = 0;

  /** Get the last known price */
  get lastPrice(): number {
    return this._lastPrice;
  }

  async connect(): Promise<void> {
    try {
      let baseUrl: string | undefined;
      if (config.binanceTestnet) {
        baseUrl = 'https://testnet.binance.vision';
      } else if (config.binanceUs) {
        baseUrl = 'https://api.binance.us';
      }

      this.client = new Binance.MainClient({
        api_key: config.binanceApiKey,
        api_secret: config.binanceApiSecret,
        baseUrl,
      });

      // Load symbol info
      await this.loadSymbolInfo();

      // Initialize simulated balances if in simulation mode
      if (config.simulationMode) {
        this.initializeSimulatedBalances();
        logger.info('Running in SIMULATION MODE - no real orders will be placed');
      }

      this.isConnected = true;
      logger.info({ testnet: config.binanceTestnet, binanceUs: config.binanceUs, simulationMode: config.simulationMode }, 'Connected to Binance');
    } catch (error) {
      logger.error({ error }, 'Failed to connect to Binance');
      throw error;
    }
  }

  private initializeSimulatedBalances(): void {
    // Initialize with configured capital
    const quoteAsset = config.quoteAsset || 'USDT';
    const baseAsset = config.baseAsset || 'DOGE';

    this.simulatedBalances.set(quoteAsset, {
      asset: quoteAsset,
      free: config.totalCapital,
      locked: 0,
      total: config.totalCapital,
    });

    this.simulatedBalances.set(baseAsset, {
      asset: baseAsset,
      free: 0,
      locked: 0,
      total: 0,
    });

    logger.info({ quoteAsset, capital: config.totalCapital }, 'Simulated balances initialized');
  }

  async disconnect(): Promise<void> {
    if (this.wsClient) {
      this.wsClient.closeAll();
      this.wsClient = null;
    }
    this.isConnected = false;
    logger.info('Disconnected from Binance');
  }

  private async loadSymbolInfo(): Promise<void> {
    if (!this.client) return;

    try {
      const exchangeInfo = await this.client.getExchangeInfo();
      const symbolData = exchangeInfo.symbols.find(
        (s) => s.symbol === config.tradingPair
      );

      if (symbolData) {
        for (const filter of symbolData.filters) {
          if (filter.filterType === 'PRICE_FILTER') {
            this.symbolInfo.tickSize = parseFloat(String((filter as { tickSize: string }).tickSize));
          } else if (filter.filterType === 'LOT_SIZE') {
            this.symbolInfo.stepSize = parseFloat(String((filter as { stepSize: string }).stepSize));
          } else if (filter.filterType === 'NOTIONAL') {
            this.symbolInfo.minNotional = parseFloat(String((filter as { minNotional: string }).minNotional));
          }
        }
        logger.debug({ symbolInfo: this.symbolInfo }, 'Symbol info loaded');
      }
    } catch (error) {
      logger.warn({ error }, 'Failed to load symbol info');
    }
  }

  roundPrice(price: number): number {
    if (!this.symbolInfo.tickSize) return price;
    return Math.floor(price / this.symbolInfo.tickSize) * this.symbolInfo.tickSize;
  }

  roundQuantity(quantity: number): number {
    if (!this.symbolInfo.stepSize) return quantity;
    return Math.floor(quantity / this.symbolInfo.stepSize) * this.symbolInfo.stepSize;
  }

  async getCurrentPrice(): Promise<number> {
    if (!this.client) throw new Error('Client not connected');

    const ticker = await this.client.getSymbolPriceTicker({
      symbol: config.tradingPair,
    });

    if (Array.isArray(ticker)) {
      throw new Error('Unexpected response from Binance');
    }

    return parseFloat(String(ticker.price));
  }

  async getBalances(): Promise<Balance[]> {
    if (!this.client) throw new Error('Client not connected');

    // Return simulated balances in simulation mode
    if (config.simulationMode) {
      return Array.from(this.simulatedBalances.values()).filter(b => b.total > 0);
    }

    const account = await this.client.getAccountInformation();
    return account.balances
      .filter((b) => parseFloat(String(b.free)) > 0 || parseFloat(String(b.locked)) > 0)
      .map((b) => ({
        asset: b.asset,
        free: parseFloat(String(b.free)),
        locked: parseFloat(String(b.locked)),
        total: parseFloat(String(b.free)) + parseFloat(String(b.locked)),
      }));
  }

  async getBalance(asset: string): Promise<Balance> {
    const balances = await this.getBalances();
    const balance = balances.find((b) => b.asset === asset);
    return balance || { asset, free: 0, locked: 0, total: 0 };
  }

  async placeLimitOrder(
    side: OrderSide,
    price: number,
    quantity: number,
    gridLevel?: number
  ): Promise<Order> {
    if (!this.client) throw new Error('Client not connected');

    const roundedPrice = this.roundPrice(price);
    const roundedQuantity = this.roundQuantity(quantity);

    // Validate minimum notional
    const notional = roundedPrice * roundedQuantity;
    if (this.symbolInfo.minNotional && notional < this.symbolInfo.minNotional) {
      throw new Error(
        `Order notional ${notional} below minimum ${this.symbolInfo.minNotional}`
      );
    }

    const clientOrderId = `grid_${gridLevel ?? 0}_${side}_${Date.now()}`;

    // SIMULATION MODE: Create simulated order instead of real one
    if (config.simulationMode) {
      return this.placeSimulatedOrder(side, roundedPrice, roundedQuantity, gridLevel, clientOrderId);
    }

    try {
      const result = await this.client.submitNewOrder({
        symbol: config.tradingPair,
        side: side as 'BUY' | 'SELL',
        type: 'LIMIT',
        timeInForce: 'GTC',
        price: roundedPrice,
        quantity: roundedQuantity,
        newClientOrderId: clientOrderId,
      });

      // Type assertion for the result
      const orderResult = result as {
        orderId: number;
        clientOrderId: string;
        symbol: string;
        side: string;
        type: string;
        price: string;
        origQty: string;
        executedQty: string;
        status: string;
        transactTime: number;
      };

      const order: Order = {
        orderId: orderResult.orderId.toString(),
        clientOrderId: orderResult.clientOrderId,
        tradingPair: orderResult.symbol,
        side: orderResult.side as OrderSide,
        orderType: orderResult.type,
        price: parseFloat(String(orderResult.price)),
        quantity: parseFloat(String(orderResult.origQty)),
        filledQuantity: parseFloat(String(orderResult.executedQty)),
        status: orderResult.status as OrderStatus,
        gridLevel,
        createdAt: new Date(orderResult.transactTime),
      };

      logger.info(
        { orderId: order.orderId, side, price: roundedPrice, quantity: roundedQuantity },
        'Order placed'
      );

      return order;
    } catch (error) {
      logger.error({ error, side, price, quantity }, 'Failed to place order');
      throw error;
    }
  }

  private placeSimulatedOrder(
    side: OrderSide,
    price: number,
    quantity: number,
    gridLevel?: number,
    clientOrderId?: string
  ): Order {
    const orderId = `SIM_${this.simulatedOrderIdCounter++}`;

    // Lock funds for the order
    const quoteAsset = config.quoteAsset || 'USDT';
    const baseAsset = config.baseAsset || 'DOGE';
    const quoteBalance = this.simulatedBalances.get(quoteAsset);
    const baseBalance = this.simulatedBalances.get(baseAsset);

    if (side === 'BUY') {
      const cost = price * quantity;
      if (quoteBalance && quoteBalance.free >= cost) {
        quoteBalance.free -= cost;
        quoteBalance.locked += cost;
      } else {
        logger.warn({ side, price, quantity, available: quoteBalance?.free }, 'Insufficient simulated balance for BUY');
      }
    } else {
      if (baseBalance && baseBalance.free >= quantity) {
        baseBalance.free -= quantity;
        baseBalance.locked += quantity;
      }
    }

    const order: SimulatedOrder = {
      orderId,
      clientOrderId: clientOrderId || `sim_${Date.now()}`,
      tradingPair: config.tradingPair,
      side,
      orderType: 'LIMIT',
      price,
      quantity,
      filledQuantity: 0,
      status: 'NEW',
      gridLevel,
      createdAt: new Date(),
      isSimulated: true,
    };

    this.simulatedOrders.set(orderId, order);

    logger.info(
      { orderId, side, price, quantity, gridLevel, simulated: true },
      '[SIMULATION] Order placed'
    );

    return order;
  }

  async cancelOrder(orderId: string): Promise<boolean> {
    if (!this.client) throw new Error('Client not connected');

    // SIMULATION MODE: Cancel simulated order
    if (config.simulationMode) {
      return this.cancelSimulatedOrder(orderId);
    }

    try {
      await this.client.cancelOrder({
        symbol: config.tradingPair,
        orderId: parseInt(orderId, 10),
      });
      logger.info({ orderId }, 'Order cancelled');
      return true;
    } catch (error) {
      logger.error({ error, orderId }, 'Failed to cancel order');
      return false;
    }
  }

  private cancelSimulatedOrder(orderId: string): boolean {
    const order = this.simulatedOrders.get(orderId);
    if (!order) return false;

    // Unlock funds
    const quoteAsset = config.quoteAsset || 'USDT';
    const baseAsset = config.baseAsset || 'DOGE';
    const quoteBalance = this.simulatedBalances.get(quoteAsset);
    const baseBalance = this.simulatedBalances.get(baseAsset);

    if (order.side === 'BUY' && quoteBalance) {
      const cost = order.price * order.quantity;
      quoteBalance.locked -= cost;
      quoteBalance.free += cost;
    } else if (order.side === 'SELL' && baseBalance) {
      baseBalance.locked -= order.quantity;
      baseBalance.free += order.quantity;
    }

    this.simulatedOrders.delete(orderId);
    logger.info({ orderId, simulated: true }, '[SIMULATION] Order cancelled');
    return true;
  }

  async cancelAllOrders(): Promise<number> {
    if (!this.client) throw new Error('Client not connected');

    // SIMULATION MODE: Cancel all simulated orders
    if (config.simulationMode) {
      const count = this.simulatedOrders.size;
      for (const orderId of this.simulatedOrders.keys()) {
        this.cancelSimulatedOrder(orderId);
      }
      logger.info({ count, simulated: true }, '[SIMULATION] All orders cancelled');
      return count;
    }

    try {
      const result = await this.client.cancelAllSymbolOrders({
        symbol: config.tradingPair,
      });
      const count = Array.isArray(result) ? result.length : 0;
      logger.info({ count }, 'All orders cancelled');
      return count;
    } catch (error) {
      logger.error({ error }, 'Failed to cancel all orders');
      return 0;
    }
  }

  async getOpenOrders(): Promise<Order[]> {
    if (!this.client) throw new Error('Client not connected');

    // SIMULATION MODE: Return simulated open orders
    if (config.simulationMode) {
      return Array.from(this.simulatedOrders.values()).filter(
        o => o.status === 'NEW' || o.status === 'PARTIALLY_FILLED'
      );
    }

    const orders = await this.client.getOpenOrders({
      symbol: config.tradingPair,
    });

    return orders.map((o) => ({
      orderId: o.orderId.toString(),
      clientOrderId: o.clientOrderId,
      tradingPair: o.symbol,
      side: o.side as OrderSide,
      orderType: o.type,
      price: parseFloat(String(o.price)),
      quantity: parseFloat(String(o.origQty)),
      filledQuantity: parseFloat(String(o.executedQty)),
      status: o.status as OrderStatus,
      gridLevel: this.extractGridLevel(o.clientOrderId),
      createdAt: new Date(o.time),
    }));
  }

  async getRecentTrades(limit = 50): Promise<Trade[]> {
    if (!this.client) throw new Error('Client not connected');

    const trades = await this.client.getAccountTradeList({
      symbol: config.tradingPair,
      limit,
    });

    return trades.map((t) => ({
      tradeId: t.id.toString(),
      orderId: t.orderId.toString(),
      tradingPair: t.symbol,
      side: t.isBuyer ? ('BUY' as OrderSide) : ('SELL' as OrderSide),
      price: parseFloat(String(t.price)),
      quantity: parseFloat(String(t.qty)),
      commission: parseFloat(String(t.commission)),
      commissionAsset: t.commissionAsset,
      realizedPnl: 0,
      createdAt: new Date(t.time),
    }));
  }

  private extractGridLevel(clientOrderId: string): number | undefined {
    if (clientOrderId?.startsWith('grid_')) {
      const parts = clientOrderId.split('_');
      if (parts.length >= 2) {
        const level = parseInt(parts[1], 10);
        if (!isNaN(level)) return level;
      }
    }
    return undefined;
  }

  onPriceUpdate(callback: PriceCallback): void {
    this.priceCallbacks.push(callback);
  }

  onOrderUpdate(callback: OrderCallback): void {
    this.orderCallbacks.push(callback);
  }

  async startPriceStream(): Promise<void> {
    let wsBaseUrl: string | undefined;
    if (config.binanceTestnet) {
      wsBaseUrl = 'wss://testnet.binance.vision';
    } else if (config.binanceUs) {
      wsBaseUrl = 'wss://stream.binance.us:9443';
    }

    this.wsClient = new Binance.WebsocketClient({
      api_key: config.binanceApiKey,
      api_secret: config.binanceApiSecret,
      beautify: true,
      wsUrl: wsBaseUrl,
    });

    this.wsClient.on('formattedMessage', (data: unknown) => {
      const msg = data as { eventType?: string; close?: string };
      if (msg.eventType === '24hrTicker' && msg.close) {
        const price = parseFloat(msg.close);
        this._lastPrice = price;

        // In simulation mode, check if any orders should be filled
        if (config.simulationMode) {
          this.checkSimulatedOrderFills(price);
        }

        for (const callback of this.priceCallbacks) {
          callback(price);
        }
      }
    });

    this.wsClient.on('error', (error) => {
      logger.error({ error }, 'WebSocket error');
    });

    this.wsClient.subscribeSpotSymbol24hrTicker(config.tradingPair);
    logger.info('Price stream started');
  }

  async startUserStream(): Promise<void> {
    if (!this.wsClient) return;

    this.wsClient.on('formattedUserDataMessage', (data: unknown) => {
      const msg = data as {
        eventType?: string;
        orderId?: number;
        newClientOrderId?: string;
        symbol?: string;
        side?: string;
        orderType?: string;
        price?: string;
        quantity?: string;
        lastFilledQuantity?: string;
        orderStatus?: string;
        eventTime?: number;
      };

      if (msg.eventType === 'executionReport') {
        const order: Order = {
          orderId: String(msg.orderId || ''),
          clientOrderId: msg.newClientOrderId,
          tradingPair: msg.symbol || '',
          side: (msg.side as OrderSide) || 'BUY',
          orderType: msg.orderType || 'LIMIT',
          price: parseFloat(msg.price || '0'),
          quantity: parseFloat(msg.quantity || '0'),
          filledQuantity: parseFloat(msg.lastFilledQuantity || '0'),
          status: (msg.orderStatus as OrderStatus) || 'NEW',
          gridLevel: this.extractGridLevel(msg.newClientOrderId || ''),
          createdAt: new Date(msg.eventTime || Date.now()),
        };

        for (const callback of this.orderCallbacks) {
          callback(order);
        }
      }
    });

    this.wsClient.subscribeSpotUserDataStream();
    logger.info('User stream started');
  }

  get connected(): boolean {
    return this.isConnected;
  }

  /**
   * Check if any simulated orders should be filled based on current price
   * BUY orders fill when price <= order price
   * SELL orders fill when price >= order price
   */
  private checkSimulatedOrderFills(currentPrice: number): void {
    const quoteAsset = config.quoteAsset || 'USDT';
    const baseAsset = config.baseAsset || 'DOGE';
    const quoteBalance = this.simulatedBalances.get(quoteAsset);
    const baseBalance = this.simulatedBalances.get(baseAsset);

    for (const [orderId, order] of this.simulatedOrders) {
      if (order.status !== 'NEW') continue;

      let shouldFill = false;

      if (order.side === 'BUY' && currentPrice <= order.price) {
        shouldFill = true;
      } else if (order.side === 'SELL' && currentPrice >= order.price) {
        shouldFill = true;
      }

      if (shouldFill) {
        // Update order status
        order.status = 'FILLED';
        order.filledQuantity = order.quantity;

        // Update balances
        if (order.side === 'BUY') {
          // Bought: locked quote -> free base
          const cost = order.price * order.quantity;
          if (quoteBalance) {
            quoteBalance.locked -= cost;
            quoteBalance.total = quoteBalance.free + quoteBalance.locked;
          }
          if (baseBalance) {
            baseBalance.free += order.quantity;
            baseBalance.total = baseBalance.free + baseBalance.locked;
          }
        } else {
          // Sold: locked base -> free quote
          const revenue = order.price * order.quantity;
          if (baseBalance) {
            baseBalance.locked -= order.quantity;
            baseBalance.total = baseBalance.free + baseBalance.locked;
          }
          if (quoteBalance) {
            quoteBalance.free += revenue;
            quoteBalance.total = quoteBalance.free + quoteBalance.locked;
          }
        }

        logger.info(
          {
            orderId,
            side: order.side,
            price: order.price,
            quantity: order.quantity,
            currentPrice,
            gridLevel: order.gridLevel,
            simulated: true,
          },
          '[SIMULATION] Order FILLED'
        );

        // Notify callbacks about the filled order
        for (const callback of this.orderCallbacks) {
          callback(order);
        }
      }
    }
  }

  /**
   * Get simulated trades (filled orders)
   */
  getSimulatedTrades(): Order[] {
    return Array.from(this.simulatedOrders.values()).filter(o => o.status === 'FILLED');
  }

  /**
   * Get current simulated balances
   */
  getSimulatedBalances(): Balance[] {
    return Array.from(this.simulatedBalances.values());
  }
}

export const binanceClient = new BinanceClient();
